compiler/fir/resolve/src/org/jetbrains/kotlin/fir/types/ConeTypeCompatibilityChecker.kt

```diff
@@ -0,0 +1,489 @@
+        typeParameterOwner: FirClassLikeDeclaration<*>,
+        parameter: FirTypeParameterRef,
+        boundTypeArgument: BoundTypeArgument,
+        compatibilityUpperBound: Compatibility,
+    ) {
+        computeIfAbsent(parameter) {
+            // the semantic of type parameter in Enum and KClass are fixed: values of types with incompatible type parameters are always
+            // incompatible.
+            val compatibilityUpperBoundForTypeArg =
+                if ((ctx.prohibitComparisonOfIncompatibleEnums && typeParameterOwner.symbol.classId == StandardClassIds.Enum) ||
+                    (ctx.prohibitComparisonOfIncompatibleClasses && typeParameterOwner.symbol.classId == StandardClassIds.KClass)
+                ) {
+                    compatibilityUpperBound
+                } else {
+                    Compatibility.SOFT_INCOMPATIBLE
+                }
+            BoundTypeArguments(mutableSetOf(), mutableSetOf(), compatibilityUpperBoundForTypeArg)
+        }.let {
+            val type = boundTypeArgument.type
+            if (boundTypeArgument.variance.allowsInPosition) {
+                it.lower += type.collectLowerBounds()
+            }
+            if (boundTypeArgument.variance.allowsOutPosition) {
+                it.upper += type.collectUpperBounds()
+            }
+        }
+    }
+
+    private fun FirClassLikeDeclaration<*>.getSuperTypes(): List<ConeClassLikeType> {
+        return when (this) {
+            is FirTypeAlias -> listOfNotNull(expandedTypeRef.coneTypeSafe())
+            is FirClass<*> -> superTypeRefs.mapNotNull { it.coneTypeSafe() }
+            else -> emptyList()
+        }
+    }
+
+    private fun ConeClassLikeType.getClassLikeElement(ctx: ConeInferenceContext): FirClassLikeDeclaration<*>? =
+        ctx.symbolProvider.getSymbolByLookupTag(lookupTag)?.fir
+
+    private fun FirClassLikeDeclaration<*>.getTypeParameter(index: Int): FirTypeParameterRef? {
+        return when (this) {
+            is FirTypeAlias -> typeParameters[index]
+            is FirClass<*> -> typeParameters[index]
+            else -> return null
+        }
+    }
+
+    /** A class declaration and the arguments bound to the declared type parameters. */
+    private data class TypeArgumentMapping(
+        val typeParameterOwner: FirClassLikeDeclaration<*>,
+        val mapping: Map<FirTypeParameterRef, BoundTypeArgument>
+    )
+
+    /** A single bound type argument to a type parameter declared in a class. */
+    private data class BoundTypeArgument(val type: ConeKotlinType, val variance: Variance)
+
+    /** Accumulated type arguments bound to a type parameter declared in a class. */
+    private data class BoundTypeArguments(
+        val upper: MutableSet<ConeClassLikeType>,
+        val lower: MutableSet<ConeClassLikeType>,
+        val compatibilityUpperBound: Compatibility
+    )
+
+    private fun ConeClassLikeType.toConcreteClass(ctx: ConeInferenceContext): ConcreteClass? {
```

Comment:
In `FirHelpers.kt` we have `toRegularClass` function. I suggest to write the similar `toClass` function and use it here.

Reply:
Sounds good! Will do!